dask/array/core.py

```diff
@@ -2062,7 +2062,7 @@ def asarray(array):
     """
     if not isinstance(array, Array):
         name = 'asarray-' + tokenize(array)
-        if not hasattr(array, 'shape'):
+        if isinstance(getattr(array, 'shape', None), Iterable):
```

Comment:
I would just stick with my original logic here, but it looks like that must have caused issues for #1151?